### PR TITLE
mesh_unweld_edges, mesh_insert_vertex_on_edge and mesh_substitute_vertex_in_faces

### PR DESCRIPTION
**Pull request by robin-oval**
three operations to insert, substitute and unweld

### CODE REVIEW

File: src/compas/datastructures/mesh/operations/insert.py [hunk 1]
```diff
@@ -3,12 +3,69 @@
...
+	----------
+	u: hashable
+		The first edge vertex.
+	v: hashable
+		The second edge vertex.
+	vkey: hashable, optional
+		The vertex key to insert.
+		Default is add a new vertex at mid-edge.
+
+	"""
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/operations/insert.py]: the behaviour of the function is not entirely clear from the docstring.
also, should it not return the key of the vertex that was inserted?
what happens if the provided vertex key already exists?
...?

File: src/compas/datastructures/mesh/operations/substitute.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	----------
+	old_vkey : hashable
+		The old vertex key.
+	new_vkey : hashable
+		The new vertex key.
+	fkeys : list, optional
+		List of face keys where to subsitute the old vertex by the new one.
+		Default is to subsitute in all faces.
+
+	"""
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/operations/substitute.py]: should the function not return the new fkeys?

File: src/compas/datastructures/mesh/operations/weld.py [hunk 3]
```diff
@@ -79,6 +85,55 @@ def mesh_unweld_vertices(mesh, fkey, where=None):
 
     return face
 
+
+def mesh_unweld_edges(mesh, edges):
+    """Unwelds a mesh along edges.
+
+    Parameters
+    ----------
+    mesh : Mesh
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/operations/weld.py:60]: description of the parameter

File: src/compas/datastructures/mesh/operations/weld.py [hunk 4]
```diff
@@ -79,6 +85,55 @@ def mesh_unweld_vertices(mesh, fkey, where=None):
...
+            if i not in edge_vertices:
+                adjacency[i] = {}
+        
+        # collect the disconnected parts around the vertex due to unwelding
+        vertex_changes[vkey] = [[new_to_old[key] for key in part] for part in connected_components(adjacency)]
+        
+    for vkey, changes in vertex_changes.items():
+        # for each disconnected part replace the vertex by a new vertex in the faces of the part
+        for change in changes:
+            mesh_substitute_vertex_in_faces(mesh, vkey, mesh.add_vertex(attr_dict = mesh.vertex[vkey]), change)
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/operations/weld.py]: remove spaces around "="

**Review by tomvanmele — CHANGES_REQUESTED** [review]
without tests it is difficult to say if the functions work as they should...

File: src/compas/datastructures/mesh/operations/substitute.py (hunk 2 above)
Comment (robin-oval) [reply src/compas/datastructures/mesh/operations/substitute.py]: added as output

File: src/compas/datastructures/mesh/operations/insert.py (hunk 1 above)
Comment (robin-oval) [reply src/compas/datastructures/mesh/operations/insert.py]: the description is extended and the key is added as output

File: src/compas/datastructures/mesh/operations/insert.py [hunk 5]
```diff
@@ -71,9 +72,91 @@ def mesh_add_vertex_to_face_edge(mesh, key, fkey, v):
         del mesh.edgedata[v, u]
 
 
+def mesh_insert_vertex_on_edge(mesh, u, v, vkey=None):
+    """Insert a vertex in the faces adjacent to an edge, between the two edge vertices.
+    If no vertex key is specified or if the key does not exist yet, a vertex is added and located at the edge midpoint.
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/operations/insert.py:74]: leave a blank line between the one-liner and the rest of the docstring...

File: src/compas/datastructures/mesh/operations/insert.py [hunk 6]
```diff
@@ -71,9 +72,91 @@ def mesh_add_vertex_to_face_edge(mesh, key, fkey, v):
...
+  
+  
 # ==============================================================================
 # Main
 # ==============================================================================
 
 if __name__ == "__main__":
-    pass
+
+    import compas
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/operations/insert.py]: we decided to not add stuff like this to the code anymore and instead just add it to the examples in the docs.
can you please update.
replace by `pass` for now...

File: src/compas/datastructures/mesh/operations/substitute.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        mesh.add_face(face_vertices, fkey)
+
+    return fkeys
+
+
+# ==============================================================================
+# Main
+# ==============================================================================
+
+if __name__ == "__main__":
```
Comment (tomvanmele) [inline src/compas/datastructures/mesh/operations/substitute.py:66]: same here.
make into an example and add to the docs...

**Comment by tomvanmele** [comment]
a few small changes and then we merge...

**Comment by robin-oval** [comment]
examples moved to docs

**Pull request merged**